Disable lock and group buttons when selected item is not CanvasItem

jsjtxietian opened the pull request:
An idea of fix https://github.com/godotengine/godot/issues/88939 , should be merged after https://github.com/godotengine/godot/pull/89132 is merged.

The origin problem lies in the fact that ParallaxBackground is not a CanvasItem so the code did not do anything special to it, I guess if this need to be fixed, we can either hide the button or make it disable. 

This pr disables lock and group buttons when there's not any selected CanvasItem. 

<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

Mickeon commented:
So what about Node3Ds?

KoBeWi commented on editor/plugins/canvas_item_editor_plugin.cpp:
```diff
@@ -3817,12 +3818,16 @@ void CanvasItemEditor::_draw_viewport() {
 		}
 	}
 
+	force_disable_lock_group = selection.size() == 1 && Object::cast_to<CanvasItem>(selection[0]) == nullptr;
```
The button will be always enabled when you select more nodes. You should check whether there is any CanvasItem in the selection.

jsjtxietian replied:
> when you select more nodes

Sorry I don't quite understand your suggestion. In this case `force_disable_lock_group` will be false and all buttons will act like before (before this pr is introduced).

KoBeWi replied:
Yep, which means that you can select e.g. 2 plain Nodes and the button will be enabled and do nothing. So right now the PR only fixes a case for single selected node.

jsjtxietian replied:
I see, thanks!

jsjtxietian committed:
Disable lock and group buttons when selected item can't be locked or grouped

jsjtxietian commented:
I've rebased the code onto master, and added back `_update_lock_and_group_button` in `_draw_viewport`, to support update the lock button immediately after click it instead of next `selection_changed`.

KoBeWi commented on editor/plugins/canvas_item_editor_plugin.cpp:
```diff
@@ -3827,6 +3831,8 @@ void CanvasItemEditor::_draw_viewport() {
 	transform.columns[2] = -view_offset * zoom;
 	EditorNode::get_singleton()->get_scene_root()->set_global_canvas_transform(transform);
 
+	_update_lock_and_group_button();
```
You could instead update when the button was pressed.

jsjtxietian replied:
But that means add update to four different places (lock, unlock, group, ungroup), and deal with UndoRedo carefully, IMHO it doesn't seem very neat

KoBeWi replied:
And the alternative is updating the buttons every time the viewport redraws, for no reason, because most actions don't affect selection.

There are signals `item_lock_status_changed` and `item_group_status_changed`. You can just connect them to update buttons.

jsjtxietian replied:
Done

KoBeWi commented on editor/plugins/canvas_item_editor_plugin.cpp:
```diff
@@ -2665,11 +2667,13 @@ void CanvasItemEditor::_update_lock_and_group_button() {
 
 	lock_button->set_visible(!all_locked);
-	lock_button->set_disabled(selection.is_empty());
+	lock_button->set_disabled(!has_canvas_item || selection.is_empty());
```
```suggestion
	lock_button->set_disabled(!has_canvas_item);
```
If selection is empty, `has_canvas_item` will always be false.

jsjtxietian replied:
Thanks！

KoBeWi commented on editor/plugins/node_3d_editor_plugin.cpp:
```diff
@@ -7365,26 +7366,31 @@ void Node3DEditor::_refresh_menu_icons() {
 
 	tool_button[TOOL_LOCK_SELECTED]->set_visible(!all_locked);
-	tool_button[TOOL_LOCK_SELECTED]->set_disabled(selection.is_empty());
+	tool_button[TOOL_LOCK_SELECTED]->set_disabled(!has_node3d_item || selection.is_empty());
```
Same here.

KoBeWi commented:
When you click non-CanvasItem node, the lock button opens:

https://github.com/godotengine/godot/assets/2223172/6e8f495a-b66e-446b-a2fb-8a9769258ee3

It's inconsistent with empty selection, which is closed, and also suggests that non-CanvasItems are "locked". Probably worth fixing.

KoBeWi commented on editor/plugins/canvas_item_editor_plugin.cpp:
```diff
@@ -2657,19 +2658,24 @@ void CanvasItemEditor::_update_lock_and_group_button() {
 		}
 	}
 
+	all_locked = all_locked && has_canvas_item;
```
`all_group` needs the same treatment.

jsjtxietian replied:
Done.

akien-mga commented:
Thanks!
